### PR TITLE
fundchannel: add the possibility to fund a channel with specific utxos

### LINKED ISSUE #2682: Feature: open channel with specific UTXO
In addition to `all` it would be nice if the `satoshi` argument for `fundchannel` can be a specific UTXO.

### PR DESCRIPTION
**Pull request by darosior**
Fixes #2682 .

This PR adds the possibility to specify a list of outpoints (`[\"txid:vout\", \"txid:vout\"]`) as the second parameter of the `fundchannel` call in addition to `'all'` or an amount in satoshis.

### CODE REVIEW

File: wallet/wallet.c [hunk 1]
```diff
@@ -377,6 +378,43 @@ const struct utxo **wallet_select_coins(const tal_t *ctx, struct wallet *w,
...
+
+		for (j = 0; j < tal_count(available); j++) {
+			struct utxo *u = tal_steal(utxos, available[j]);
+
+			if (bitcoin_txid_eq(&u->txid, txid) && u->outnum == output_index) {
+				    tal_arr_expand(&utxos, u);
+			    if (!wallet_update_output_status(
+				    w, &available[j]->txid, available[j]->outnum,
+				    output_state_available, output_state_reserved))
+				    fatal("Unable to reserve output");
```
Comment (rustyrussell) [inline wallet/wallet.c]: Um, does this mean I crash the daemon if I try to fund two channels at once from the same UTXO?

File: common/wallet_tx.c [hunk 2]
```diff
@@ -39,6 +46,37 @@ struct command_result *param_wtx(struct command *cmd,
...
+				 const char *buffer,
+				 const jsmntok_t *tok,
+				 struct wallet_tx *wtx)
+{
+	size_t i;
+	const jsmntok_t *curr;
+	const char **outpoints = tal_arr(cmd, const char*, 0);
+
+	json_for_each_arr(i, curr, tok)
+		tal_arr_expand(&outpoints, json_strdup(tmpctx, buffer, curr));
```
Comment (rustyrussell) [inline common/wallet_tx.c]: You want to use `outputs` and the context for json_strdup here: it's a nice convention that a container owns all its members.

File: common/wallet_tx.c [hunk 3]
```diff
@@ -39,6 +46,37 @@ struct command_result *param_wtx(struct command *cmd,
...
+	if (!tal_count(outpoints))
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+			"Could not decode any of the outpoints. The utxos should be"
+			" specified as an array of 'txid:output_index'.");
```
Comment (rustyrussell) [inline common/wallet_tx.c]: Empty array check should probably be left to later.

File: common/wallet_tx.c (hunk 2 above)
Comment (rustyrussell) [reply common/wallet_tx.c]: In `listfunds` we use `txid` and `output` fields.  Since we have JSON anyway, I suggest sticking with that, rather than using a boutique encoding for outpoints?

File: common/wallet_tx.c [hunk 4]
```diff
@@ -39,6 +46,37 @@ struct command_result *param_wtx(struct command *cmd,
...
+
+	if (!wtx->utxos)
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+			"Could not decode all of the outpoints. The utxos should be"
+			" specified as an array of 'txid:output_index'.");
+	if (tal_count(wtx->utxos) == 0)
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				"No matching utxo was found from the wallet."
+				"You can get a list of the wallet utxos with"
+				" the `listfunds` RPC call.");
```
Comment (rustyrussell) [inline common/wallet_tx.c]: Parsing here, before asking wallet_select_specific(), makes this code path simpler too.  If you can't decode txid or outnum, you give a meaningful error here.

Just hand wallet_select_specific an array of txids and an array of outnums: a failure means some outpoint wasn't available.

File: wallet/wallet.c [hunk 5]
```diff
@@ -377,6 +378,43 @@ const struct utxo **wallet_select_coins(const tal_t *ctx, struct wallet *w,
...
+	size_t i, j;
+	char **outpoint;
+	unsigned int output_index;
+	struct bitcoin_txid *txid;
+	struct utxo **available = wallet_get_utxos(ctx, w, output_state_available);
+	const struct utxo **utxos = tal_arr(ctx, const struct utxo*, 0);
+	tal_add_destructor2(utxos, destroy_utxos, w);
+
+	for (i = 0; i < tal_count(outpoints); i++)
+	{
```
Comment (rustyrussell) [inline wallet/wallet.c]: style nit: `{` belongs on line above.

File: wallet/wallet.c [hunk 6]
```diff
@@ -377,6 +378,43 @@ const struct utxo **wallet_select_coins(const tal_t *ctx, struct wallet *w,
...
 }
 
+const struct utxo **wallet_select_specific(const tal_t *ctx, struct wallet *w,
+					const char **outpoints)
+{
+	size_t i, j;
+	char **outpoint;
+	unsigned int output_index;
+	struct bitcoin_txid *txid;
+	struct utxo **available = wallet_get_utxos(ctx, w, output_state_available);
```
Comment (rustyrussell) [inline wallet/wallet.c]: available should either be allocated off NULL (which by convention, indicates you're explicitly going to take/free it, as you do below), or just off tmpctx and forget about it.

File: wallet/wallet.c [hunk 7]
```diff
@@ -377,6 +378,43 @@ const struct utxo **wallet_select_coins(const tal_t *ctx, struct wallet *w,
...
+		if (!bitcoin_txid_from_hex(outpoint[0], strlen(outpoint[0]), txid)
+				|| opt_set_uintval(outpoint[1], &output_index))
+			return NULL;
+
+		for (j = 0; j < tal_count(available); j++) {
+			struct utxo *u = tal_steal(utxos, available[j]);
+
+			if (bitcoin_txid_eq(&u->txid, txid) && u->outnum == output_index) {
+				    tal_arr_expand(&utxos, u);
+			    if (!wallet_update_output_status(
```
Comment (rustyrussell) [inline wallet/wallet.c]: Weird indent?

File: tests/test_connection.py [hunk 8]
```diff
@@ -753,6 +753,20 @@ def test_funding_toolarge(node_factory, bitcoind):
...
+    l1, l2 = node_factory.line_graph(2, fundchannel=False)
+
+    # Get 2 differents utxo
+    l1.fundwallet(0.01 * 10**8)
+    l1.fundwallet(0.01 * 10**8)
+    wait_for(lambda: len(l1.rpc.listfunds()['outputs']) == 2)
+
+    # Then fund a channel with them
+    utxos = [utxo["txid"] + ":" + str(utxo["output"]) for utxo in l1.rpc.listfunds()["outputs"]]
+    l1.rpc.fundchannel(l2.info["id"], utxos)
```
Comment (rustyrussell) [inline tests/test_connection.py]: OK, test that you can't grab some bogus UTXO too?  And maybe an already-consumed UTXO?

File: wallet/wallet.c (hunk 1 above)
Comment (darosior) [reply wallet/wallet.c]: I think so, as in the already used function to select outputs : https://github.com/ElementsProject/lightning/blob/35827111ee837fc0118e0190b9e12539adae81b3/wallet/wallet.c#L315-L318
Anyway we consume the whole utxos if they are directly passed to `fundchannel` so one could not fund 2 channels, or am I missing something ?

File: common/wallet_tx.c (hunk 2 above)
Comment (darosior) [reply common/wallet_tx.c]: > In listfunds we use txid and output fields. Since we have JSON anyway, I suggest sticking with that, rather than using a boutique encoding for outpoints?

Does it mean that the user should pass an array of objects ? `[{"txid":"....", "output":0}, ...]`

File: tests/test_connection.py (hunk 8 above)
Comment (darosior) [reply tests/test_connection.py]: Ok, how could I get a bogus UTXO ?

File: wallet/wallet.c (hunk 1 above)
Comment (ZmnSCPxj) [reply wallet/wallet.c]: I agree with @darosior, line 388 of the new code is `wallet_get_utxos` which ensures that we can only change update status of UTXOs with `output_state_available`.  Will mark as resolved.

File: tests/test_connection.py (hunk 8 above)
Comment (ZmnSCPxj) [reply tests/test_connection.py]: Dump 32 bytes from `/dev/random` for txid plus 2 bytes for outnum?

**Comment by darosior** [comment]
@rustyrussell thank you for the review. I rebased to apply most of the requested changes : 
- Parse outpoints before calling the utxo selection function
- Add some test for failing cases

Added a CHANGELOG brag line :)

File: tests/test_connection.py (hunk 8 above)
Comment (darosior) [reply tests/test_connection.py]: I didn't see your comment and finally added a test that makes a node funding a channel with utxos it doesn't own, is it OK ? It takes the same code path.

Comment (ZmnSCPxj) [reply tests/test_connection.py]: It is fine.

File: common/wallet_tx.c [hunk 9]
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
...
+				 const char *name,
+				 const char *buffer,
+				 const jsmntok_t *tok,
+				 struct wallet_tx *wtx)
+{
+	size_t i;
+	const jsmntok_t *curr;
+	struct bitcoin_txid **txids;
+	unsigned int **outnums;
+	char **outpoint = tal_arr(tmpctx, char*, 0);
```
Comment (niftynei) [inline common/wallet_tx.c]: nit: it's weird to me that you allocate this one inline with the declaration, but the others (`txids` and `outnums`) below.

File: common/wallet_tx.c [hunk 10]
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
...
+
+	txids = tal_arr(cmd, struct bitcoin_txid*, 0);
+	outnums = tal_arr(cmd, unsigned int*, 0);
+	json_for_each_arr(i, curr, tok) {
+		outpoint = tal_strsplit(cmd, json_strdup(cmd, buffer, curr), ":",
+				STR_NO_EMPTY);
```
Comment (niftynei) [inline common/wallet_tx.c]: nit: you should line up the beginning of this with the end of the `(` above

```
tal_strsplit(cmd ..
             STR_NO_EMPTY
```

File: common/wallet_tx.c [hunk 11]
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
...
+		if (tal_count(outpoint) < 3 || strlen(outpoint[0]) != 64)
```
Comment (niftynei) [inline common/wallet_tx.c]: Why are you expecting 'outpoint' to be at least 3 here? Why not check that it's exactly 3? Moving 3 and 64 to constants might be a better way of expressing what you're checking as well.

This would be easier to review if you had included an example of the expected rpc call in the commit notes.

File: common/wallet_tx.c [hunk 12]
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
...
+			break;
+
+		struct bitcoin_txid *txid = tal(tmpctx, struct bitcoin_txid);
+		unsigned int *outnum = tal(tmpctx, unsigned int);
```
Comment (niftynei) [inline common/wallet_tx.c]: It's weird to see `unsigned int`s allocated, but maybe I'm just new to this.

You probably want to allocate these off of the same thing as the array that you're putting them in, either `cmd` or `outnums`.

File: common/wallet_tx.c [hunk 13]
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
...
+		if (!bitcoin_txid_from_hex(outpoint[0], strlen(outpoint[0]), txid)
+				|| opt_set_uintval(outpoint[1], outnum))
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				"Could not decode the outpoint with txid \"%s\" and "
```
Comment (niftynei) [inline common/wallet_tx.c]: nit: should be aligned with `(` in previous line.

File: common/wallet_tx.c [hunk 14]
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
...
+		tal_arr_expand(&outnums, outnum);
+	}
+
+	if (!tal_count(txids) || !tal_count(outnums))
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+			"Please specify an amount in satoshis, 'all' or an array"
+			" of 'txid:output_index'.");
+
+	wtx->utxos = wallet_select_specific(cmd, cmd->ld->wallet, txids, outnums);
+	tal_free(txids);
```
Comment (niftynei) [inline common/wallet_tx.c]: if the `command_fail`s at any point previous to this, do we need to call `tal_free` for each of those branches as well?

File: common/wallet_tx.c [hunk 15]
```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
...
+	tal_free(outnums);
+
+	if (!wtx->utxos)
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+			"Could not decode all of the outpoints. The utxos should be"
+			" specified as an array of 'txid:output_index'.");
+	if (tal_count(wtx->utxos) == 0)
```
Comment (niftynei) [inline common/wallet_tx.c]: nit: you can also do `!tal_count(wtx->utxos)`

File: common/wallet_tx.c [hunk 16]
```diff
@@ -104,3 +166,34 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
...
+					u32 fee_rate_per_kw,
+					size_t out_len)
+{
+	struct command_result *res;
+	size_t weight, i;
+
+	tx->amount = AMOUNT_SAT(0);
+	/* The transaction has `tal_count(tx.utxos)` inputs and one output */
+	/* version + in count + out count + locktime    index + value + script length */
+	weight = 4 * (4 + 1 + 1 + 4) + 4 * (8 + 1 + out_len);
```
Comment (niftynei) [inline common/wallet_tx.c]: Technically the `out_len` might be longer than `1` but this is probably fine.

```
(8 + 1 + out_len)
```

File: lightningd/opening_control.c [hunk 17]
```diff
@@ -1059,8 +1059,15 @@ static struct command_result *json_fund_channel(struct command *cmd,
...
-			       BITCOIN_SCRIPTPUBKEY_P2WSH_LEN, maxheight);
+	if (tal_count(fc->wtx.utxos)) {
+		/* We consume the whole specified utxo(s) */
+		fc->wtx.change = AMOUNT_SAT(0);
+		fc->wtx.change_key_index = 0;
+		res = wtx_fees_from_utxos(&fc->wtx, *feerate_per_kw,
+				BITCOIN_SCRIPTPUBKEY_P2WSH_LEN);
+	} else
+		res = wtx_select_utxos(&fc->wtx, *feerate_per_kw,
+				BITCOIN_SCRIPTPUBKEY_P2WSH_LEN, maxheight);
```
Comment (niftynei) [inline lightningd/opening_control.c]: nit: line up with `(`

File: common/wallet_tx.c [hunk 18]
```diff
@@ -104,3 +166,34 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
...
+	for (i = 0; i < tal_count(tx->utxos); ++i) {
```
Comment (niftynei) [inline common/wallet_tx.c]: nit: use postfix, not prefix `++i` -> `i++`

File: CHANGELOG.md [hunk 19]
```diff
@@ -25,6 +25,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Config: Increased default amount for minimal channel capacity from 1k sat to 10k sat.
 - Build: Non-developer builds are now done with "-Og" optimization.
 - JSON API: `pay` will no longer return failure until it is no longer retrying; previously it could "timeout" but still make the payment.
+- JSON API: `fundchannel` now accepts an utxo array to use to fund the channel as the second parameter (in addition to 'all' or an amount in sats)
```
Comment (niftynei) [inline CHANGELOG.md]: nit: instead of saying `in addition to`, I'd reword this to `instead of` as you can only provide one of the three ('all', satoshis, or a utxo set)

File: common/wallet_tx.c (hunk 15 above)
Comment (rustyrussell) [reply common/wallet_tx.c]: ... I prefer == 0 here, since it's literally a count.

**Comment by rustyrussell** [comment]
Agree with all @niftynei feedback.  In particular, I think adding a separate optional 'utxos' param is nicer: if not specified, utxos are selected from current wallet utxos as now.  That does allow us to have a change output from UTXOs if we want to (or 'all' if we don't).

File: common/wallet_tx.c (hunk 9 above)
Comment (darosior) [reply common/wallet_tx.c]: It seemed weird to me too : I was not aware of the convention @rustyrussell pointed out in his review (to not directly allocate a pointer I will free). Since I dont free this one (`tmpctx`) I thought it was ok to allocate it directly.
I have to admit I'm a little bit confused with which style to use each time I declare a variable.

File: common/wallet_tx.c (hunk 11 above)
Comment (darosior) [reply common/wallet_tx.c]: I test the size of the outpoint array (which always ends with NULL and must have 2elements). I put this condition while testing and should have changed it.

File: common/wallet_tx.c (hunk 12 above)
Comment (darosior) [reply common/wallet_tx.c]: I allocated them off `tmpctx` because they are temo variable, however I'm also new to this and am not sure it's the correct way.

File: common/wallet_tx.c (hunk 14 above)
Comment (darosior) [reply common/wallet_tx.c]: No because tal allocated variables passed to this function are off `cmd`, which is desallocated on `comand_error` and then so they are

File: common/wallet_tx.c (hunk 16 above)
Comment (darosior) [reply common/wallet_tx.c]: That's why it's a variable

File: CHANGELOG.md (hunk 19 above)
Comment (darosior) [reply CHANGELOG.md]: Ok. I appreciate that you review my english too, since I'm not a native speaker.

**Comment by darosior** [comment]
> Lots of little things, nice work overall!
> 
> I think the biggest question is whether or not we want to have this in addition to the 'amount' field. It seems like it'd be nice to be able to specify a utxo set and a total amount to fund as well, which isn't possible with the current design.
> 
> How you do you see people using this @darosior ?

Thank you for the review.
I see people using it in a few corner cases : a traditional user doing `newaddr`, `connect`, and `fundchannel` won't use it at all. Some users having a lot of tiny inputs wanting more control over them could use it. Actually I think most people who'll use it will consume the whole utxo, that's why I made this `command_result` of `command_result` in case the parameter is an array : I thought making another parameter will just make them copy paste the utxo amount.

File: common/wallet_tx.c (hunk 11 above)
Comment (niftynei) [reply common/wallet_tx.c]: Ok cool. Moving to JSON format like rusty recommended above will probably change this too :)

**Comment by rustyrussell** [comment]
Ideally, we'd allow them to specify any combination, but that's a bit ambitious.  I really think a "utxos" arg is best here.

This will combine well with #2713 too

**Comment by darosior** [comment]
Ok, let's do this

File: common/wallet_tx.c (hunk 11 above)
Comment (darosior) [reply common/wallet_tx.c]: I don't think, since I need to decode the txid and the outpoint out of the jmn token

Comment (darosior) [reply common/wallet_tx.c]: Ok, finally made a json_helper for json token to transaction id :D

Comment (darosior) [reply common/wallet_tx.c]: Oh, I just saw that you did the same in your external wallet funding https://github.com/ElementsProject/lightning/pull/2672/commits/d8dcbce4485b91a9b4e6f9037270fedf18b19002

**Comment by rustyrussell** [comment]
Sorry, you really can't do the API this way.  As @niftynei points out, you may well want to specify both utxos and the amount to spend, so it needs to be a separate 'utxos' argument.

And while txid:outnum is neat on the commandline, it's nasty JSON.  You can always create a plugin which does converts from this syntax :)

**Comment by darosior** [comment]
> Sorry, you really can't do the API this way. As @niftynei points out, you may well want to specify both utxos and the amount to spend, so it needs to be a separate 'utxos' argument.

When saying "let's do this" I meant "ok I'll work on the separate 'utxos' argument version (I was not pushing the initial implementation without argument ^^), it a translated sentence from french and has apparently not the same signification in english.

> And while txid:outnum is neat on the commandline, it's nasty JSON. You can always create a plugin which does converts from this syntax :)

I just saw your comment now that I finished to work on the new implementation. I finally kept using JSON as you requested (I added a bitcoin_txid json helper), relying on the fact that a txid would always be 64 hex characters long.

File: common/wallet_tx.c [hunk 20]
```diff
@@ -39,6 +43,63 @@ struct command_result *param_wtx(struct command *cmd,
...
+	struct bitcoin_txid **txids = tal_arr(cmd, struct bitcoin_txid*, 0);
+	unsigned int **outnums = tal_arr(cmd, unsigned int*, 0);
+
+	json_for_each_arr(i, curr, tok) {
+		/* A txid is always 32 bytes (64 hex characters) long */
+        const jsmntok_t txid_tok = {JSMN_STRING, curr->start,
+									curr->start + 64, curr->size};
+		/* Pass the `:` separator */
+		const jsmntok_t outnum_tok = {JSMN_STRING, txid_tok.end + 1,
+									curr->end, curr->size};
```
Comment (rustyrussell) [inline common/wallet_tx.c]: This code can read past the end of malformed input. 

Prefer something like this:

```
static bool split_tok(const char *buffer, const jsmntok_t *t, char split, jsmntok_t *a, jsmntok_t *b)
{
        const char *p = memchr(buffer + t->start, split, buffer + t->end);

        if (!p)
                 return false;

        *a = *b = *t;
        a->end = p - 1 - buffer;
        b->start = p + 1 - buffer;
        return true;
}
```

File: common/wallet_tx.c [hunk 21]
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
...
+
+	tx->change = AMOUNT_SAT(0);
+	tx->change_key_index = 0;
+	total_amount = AMOUNT_SAT(0);
+
+	/* The transaction has `tal_count(tx.utxos)` inputs and one output */
+	/* (version + in count + out count + locktime)  (index + value + script length) */
+	weight = 4 * (4 + 1 + 1 + 4) + 4 * (8 + 1 + out_len);
+	for (i = 0; i < tal_count(utxos); i++) {
+        if (!*utxos[i]->blockheight || *utxos[i]->blockheight > maxheight) {
```
Comment (rustyrussell) [inline common/wallet_tx.c]: Weird indent?

File: common/wallet_tx.c [hunk 22]
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
...
+
+	fee_estimate = amount_tx_fee(fee_rate_per_kw, weight);
+
+	if (tx->all_funds || amount_sat_eq(tx->change, AMOUNT_SAT(0))) {
+		tx->amount = total_amount;
+        if (!amount_sat_sub(&tx->amount, tx->amount, fee_estimate))
+			return command_fail(tx->cmd, FUND_CANNOT_AFFORD,
+								"Cannot afford transaction");
+	} else {
+        if (!amount_sat_sub(&tx->change, tx->change, fee_estimate))
```
Comment (rustyrussell) [inline common/wallet_tx.c]: Weird indent again...

File: common/wallet_tx.c [hunk 23]
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
...
+		weight += (8 + 1 + out_len) * 4;
+	}
+
+	fee_estimate = amount_tx_fee(fee_rate_per_kw, weight);
+
+	if (tx->all_funds || amount_sat_eq(tx->change, AMOUNT_SAT(0))) {
+		tx->amount = total_amount;
+        if (!amount_sat_sub(&tx->amount, tx->amount, fee_estimate))
+			return command_fail(tx->cmd, FUND_CANNOT_AFFORD,
+								"Cannot afford transaction");
```
Comment (rustyrussell) [inline common/wallet_tx.c]: Maybe give a clue here as to what the fee estimate was?

File: common/wallet_tx.c [hunk 24]
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
...
+	} else {
+        if (!amount_sat_sub(&tx->change, tx->change, fee_estimate))
+            return command_fail(tx->cmd, FUND_CANNOT_AFFORD,
+								"Cannot afford transaction");
```
Comment (rustyrussell) [inline common/wallet_tx.c]: Fee estimate printing here too?

File: common/wallet_tx.c [hunk 25]
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
...
+	struct amount_sat total_amount, fee_estimate;
+
+	tx->change = AMOUNT_SAT(0);
+	tx->change_key_index = 0;
+	total_amount = AMOUNT_SAT(0);
+
+	/* The transaction has `tal_count(tx.utxos)` inputs and one output */
+	/* (version + in count + out count + locktime)  (index + value + script length) */
+	weight = 4 * (4 + 1 + 1 + 4) + 4 * (8 + 1 + out_len);
+	for (i = 0; i < tal_count(utxos); i++) {
```
Comment (rustyrussell) [inline common/wallet_tx.c]: mildly prefer C-99-style `(for size_t i = 0;...` here.

File: common/wallet_tx.c [hunk 26]
```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
...
+        if (!amount_sat_sub(&tx->amount, tx->amount, fee_estimate))
+			return command_fail(tx->cmd, FUND_CANNOT_AFFORD,
+								"Cannot afford transaction");
+	} else {
+        if (!amount_sat_sub(&tx->change, tx->change, fee_estimate))
+            return command_fail(tx->cmd, FUND_CANNOT_AFFORD,
+								"Cannot afford transaction");
+	}
+	res = check_amount(tx, tx->amount);
+    return res;
```
Comment (rustyrussell) [inline common/wallet_tx.c]: indent is weird, but just `return check_amount...` instead of a temporary variable?

**Comment by rustyrussell** [comment]
OK, minor fixes only.  I'm trying to tag rc1, but there are a couple of other merges pending review, so I'll wait (was tempted to tweak it myself, but that's a bad habit!).

File: common/wallet_tx.c (hunk 20 above)
Comment (darosior) [reply common/wallet_tx.c]: That was what I had at first, but I need `const` tokens and didnt want to do multiple cast. Is it preferable to cast ?

**Comment by darosior** [comment]
@rustyrussell I corrected the minor fixes, added the split function (as a helper) and feerate estimate in errors, I also split the first check about parsing the txid and vout in two parts to have more accurate errors.

**Comment by rustyrussell** [comment]
Ack https://github.com/ElementsProject/lightning/commit/035ea5be213c235c2e17ca8b5da867adf789447a

GitHub ordered the commits weirdly, but I checked locally and they're in the expected order...
